File: 1010-pairs-of-songs-with-total-durations-divisible-by-60/1010-pairs-of-songs-with-total-durations-divisible-by-60.py
class Solution:
    
    ### 实际上就是 每次新得到一个reminder 和现存对应的60-reminder的数量配对即可
    ### 比如20 40 20 40
    def numPairsDivisibleBy60(self, time):
        c = [0] * 60
        res = 0
        for t in time:
            res += c[-t % 60]
            c[t % 60] += 1
        return res
    
    
#     def numPairsDivisibleBy60(self, time: List[int]) -> int:
#         reminders = [0 for _ in range(60)]
        
#         for t in time:
#             reminders[t%60] += 1
            
#         pairs = 0
        
#         if reminders[0] > 1:
#             k = reminders[0]
#             curr = 1
#             while k:
#                 curr *= k
#                 k -= 1
#             curr /= 2
        
#             pairs += int(curr)
                
            
#         for i in range(1, 30):
#             if reminders[i] and reminders[60-i]:
#                 curr = reminders[i] * reminders[60-i] 
#                 pairs += curr
    
#         if reminders[30] > 1:
#             k = reminders[30]
#             curr = 1
#             while k:
#                 curr *= k
#                 k -= 1
#             curr /= 2
#             pairs += int(curr)
                
                
#         return pairs
            
        

File: 1010-pairs-of-songs-with-total-durations-divisible-by-60/1010-pairs-of-songs-with-total-durations-divisible-by-60.java
class Solution {
    public int numPairsDivisibleBy60(int[] time) {
        int res = 0;
        int[]freqs = new int[60];

        for (int t : time) {
            int reminder = t % 60;
            freqs[reminder] += 1;
        }

        for (int i = 0; i < 31; i++) {
            if(i == 0 || i == 30){
                res += (freqs[i] * (freqs[i] - 1)) / 2;
            }else{
                res += (freqs[i] * freqs[60-i]);
            }
            

        }
        if(res  < 0){
             res += Integer.MAX_VALUE + 1;
        }

        return res ;
    }
}